### PR TITLE
sac implement

### PR DESCRIPTION
**Pull request by witwolf**

### CODE REVIEW

File: alf/algorithms/sac_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        info = SacCriticInfo(
+            log_pi=log_pi,
+            critic1=critic1, critic2=critic2,
+            target_critic1=target_critic1,
+            target_critic2=target_critic2)
+
+        return state, info
+
+    def _alpha_train_step(self, exp: Experience, state: SacActorState):
+        action, log_pi, actor_state = self._actions_and_log_probs(exp, state)
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: Can we avoid the duplicated computation of self._actions_and_log_probs(exp, state) as _critic_train_step, _actor_train_step all use it.

File: alf/algorithms/sac_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            value=critic_info.critic2,
+            target_value=target_q,
+            final_time_step=final_time_step,
+            final_target_value=final_target_q)
+
+        critic_loss = critic_loss1.loss + critic_loss2.loss
+        return LossInfo(loss=critic_loss, extra=critic_loss)
+
+    def _calc_alpha_loss(self, training_info):
+        alpha_info = training_info.info.alpha
```
[the remark this answers is not in the document]
Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: updated

File: alf/algorithms/sac_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+        target_q = target_q_fn(
+            critic_info.target_critic1,
+            critic_info.target_critic2,
+            critic_info.log_pi)
+
+        final_target_q = target_q_fn(
+            final_critic_info.target_critic1,
+            final_critic_info.target_critic2,
+            final_critic_info.log_pi)
```
[the remark this answers is not in the document]
Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: updated

File: alf/algorithms/sac_algorithm.py (hunk 1 above)
Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: add SacShareState for store actor state for actor, critic, alpha and calc _actions_and_log_probs only once ,but it seems not as clear as before

File: alf/algorithms/sac_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        self._actor_optimizer = actor_optimizer
+        self._critic_optimizer = critic_optimizer
+        self._alpha_optimizer = alpha_optimizer
+
+        if critic_loss is None:
+            critic_loss = OneStepTDLoss(debug_summaries=debug_summaries)
+        self._critic_loss = critic_loss
+
+        flat_action_spec = tf.nest.flatten(self._action_spec)
+        self._target_entropy = -np.sum([
```
[the remark this answers is not in the document]
Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: added

**Pull request merged**